### ISSUE TITLE
How to plot Cox model with ggsurvplot()?

### ISSUE THREAD

**Issue by kassambara** [kassambara/survminer#67]

**Comment by kassambara** [comment]
```r
require('survival')
require('survminer')
res.cox <- coxph(Surv(time, status) ~ age + sex + wt.loss, data =  lung)

# Visualize the predicted survival proportion
ggsurvplot(survfit(res.cox), color = "#2E9FDF",
           ggtheme = theme_minimal())
```

**Comment by teigentler** [comment]
I tried 
ggsurvplot(survfit(res.cox,newdata=lung), color = "#2E9FDF",
          ggtheme= theme_minimal())
for testing what happens when I fit new data to the cox model. Unfortunately survminer breaks with:
Error in rbind(deparse.level, ...) : 
  numbers of columns of arguments do not match

**Comment by kassambara** [comment]
Yes, I  have already noted that and I'm working on that.

This issue should be resolved the following R script:

```r
library("survival")
# Fit cox ph
#%%%%%%%%%%%%%
res.cox <- coxph(Surv(time, status) ~ age + sex + wt.loss, data =  lung)
res.cox

# Survival curves by sex after adjusting by age and wt.loss
#%%%%%%%%%%%%%%%%%%%%%%%%
# we construct a new data frame with two rows, 
# one for each value of sex; the other covariates are fixed to their average values 
# Create the new data  
sex_df <- with(lung,
               data.frame(sex = c(1, 2), 
                          age = rep(mean(age, na.rm = TRUE), 2),
                          wt.loss = rep(mean(wt.loss, na.rm = TRUE), 2)
                          )
               )
sex_df

# Survival curves
library("ggplot2")
require("tidyr")
ncurve <- ncol(fit$surv)
ssurv <- tidyr::gather(as.data.frame(fit$surv), "strata", "surv")
ssurv$time <- rep(fit$time, ncurve )
ggplot(data = ssurv) + geom_step(aes(time, surv, color = strata))
```
![cox-proportional-hazards-survival-coxph-2-2](https://cloud.githubusercontent.com/assets/3313355/19679038/8ea07ef2-9aa0-11e6-8627-01011c4035d6.png)

Best,
/A

**Comment by kassambara** [comment]
Now, `ggsurvplot()` can be used to plot cox model:
   
```r
library("survival")
# Fit cox ph
#%%%%%%%%%%%%%
res.cox <- coxph(Surv(time, status) ~ age + sex + wt.loss, data =  lung)
res.cox

# Survival curves by sex after adjusting by age and wt.loss
#%%%%%%%%%%%%%%%%%%%%%%%%
# we construct a new data frame with two rows, 
# one for each value of sex; the other covariates are fixed to their average values 
# Create the new data  
new_df <- with(lung,
               data.frame(sex = c(1, 2), 
                          age = rep(mean(age, na.rm = TRUE), 2),
                          wt.loss = rep(mean(wt.loss, na.rm = TRUE), 2)
               )
)
new_df

# Survival curves with new data
#%%%%%%%%%%%%%%%%%%%%%%%%%%%
library(survminer)
fit <- survfit(res.cox, newdata = new_df)
ggsurvplot(fit, conf.int = TRUE, palette = "Dark2", 
           censor = FALSE, surv.median.line = "hv")
```
![rplot](https://cloud.githubusercontent.com/assets/3313355/19835060/f68f8d22-9e7a-11e6-987e-5b1ade7a4eaa.png)

**Comment by markdanese** [comment]
I love the survminer package.  It makes things so much better, and I use it quite frequently (thanks!).  Please don't take my comments below as criticism -- they are intended to be helpful.
 
I would suggest that this option for adjusted survival curves be considered carefully.  See https://cran.r-project.org/web/packages/survival/vignettes/adjcurve.pdf which is written by the author of the survival package (specifically section 5.2, where there are other references for why this is not usually a great idea).  The survival curve at the average value of covariates is generally not recommended.  One better approach is to predict survival for all individuals in the cohort, and then take the average of the predicted curves by groups of interest (for example, sex, age group, etc.).  It would be great if the survminer package would support this.  

Below is my rough code to accomplish this.  Not nearly ready for your package, but at least it might help to illustrate what I mean.  (I did this for another purpose and stumbled across this github issue).  One challenge is that the standard errors need to be bootstrapped.

Note that there are many other ways to generate average survival curves, and the pdf referenced above goes through these options with the proper variance estimators.

```
library(survival)
library(data.table)
library(magrittr)
library(ggplot2)

cancer2 <- 
    copy(cancer) %>% 
    setDT() # dataset stored in the survival package for testing
cancer3 <- cancer2[complete.cases(cancer2)] # taking complete cases to avoid NA observations on fitted model

x <- coxph(Surv(time, status) ~ age + factor(sex) + factor(ph.ecog), data = cancer3) # cox model
pred <- survfit(x, cancer3) # predicted survival curve using raw data
timepoints <- c(0, pred$time) # time points based on study failure times, adding time = 0 for starting point
timenames <- paste0("time", timepoints) # variable names for times in the study
adj_surv <- 
    t(pred$surv) %>% 
    as.data.table() # taking the predicted survival curves for each person in the study based on the cox model
adj_surv <- cbind(time0 = 1, adj_surv) # adding survival probability = 1 at time = 0
setnames(adj_surv, timenames) # naming all of the columns
cancer4 <- cbind(cancer3, adj_surv) # put togethe with original data
cancer5 <- cancer4[, lapply(.SD, mean), .SDcols = timenames, by = sex] # key step -- calculate mean survival at each time point, by group of interest (e.g., sex)
graph <- 
    cbind(timepoints, t(cancer5[, -1])) %>% 
    as.data.table() # make data for graph
names(graph) <- c("time", "men", "women") # variable names
graph <- melt.data.table(data = graph, id.vars = "time", variable.name = "Sex", value.name = "Survival") # stack data for ggplot
ggplot(graph, aes(x = time, y = Survival, color = Sex)) +
    geom_step() +
    scale_y_continuous(limits = c(0, 1)) +
    theme_bw()
```

**Comment by kassambara** [comment]
Thank you @markdanese  for this suggestion, which will help us to improve survminer. I will work on this issue as soon as possible.

Best,
/alboukadel

**Comment by markdanese** [comment]
@kassambara -- thanks very much.  I turned it into a function and the script is available at at this gist:  https://gist.github.com/markdanese/d1ddfdb2f618373c719bdb444ca34be9 

It requires data.table but it can be done easily with base R too -- I use data.table so much that I can't really determine the best way to do it in base R.  This will at least help anyone who is trying to do this in the meantime.

**Comment by kassambara** [comment]
Thank you @markdanese for this great job! I'll see how to adapt your script and to include it in the survminer package.

**Comment by drkirmani** [comment]
Apologies for dragging up this old thread. I've been using survminer for a while - it's an absolute gem. However, I'm now getting a repeat of the old "Error in rbind(deparse.level, ...) :numbers of columns of arguments do not match" error above when trying to use ggsurvplot on a cox model. Have used your code almost verbatim above and can get it to plot(fit) but not ggsurvplot which gives the number of columns error. Any ideas if this is a new issue?

**Comment by kassambara** [comment]
There is now a dedicated function, named `ggcoxadjustedcurves`() [in the survminer] to plot adjusted survival curves. Please, read this http://www.sthda.com/english/wiki/survminer-0-3-0#adjusted-survival-curves-for-cox-model

Please, create a new issue with a reproducible example so that we can help you.

**Comment by drkirmani** [comment]
Thanks, that's great. I guess the reference material I was looking at was deprecated. Apologies. The arguments for the aesthetics aren't the same as in ggsurvplot, but if I need help with that I'll open a new issue accordingly.

**Comment by bounlu** [comment]
For me it still requires to pass the original data frame to the survfit function when plotting:

```
fit <- survfit(res.cox, newdata = new_df)
ggsurvplot(fit, conf.int = TRUE, palette = "Dark2", 
           censor = FALSE, surv.median.line = "hv")

Error in .get_data(fit, data = data, complain = FALSE) : 
  The `data` argument should be provided either to ggsurvfit or survfit.


fit <- survfit(res.cox, newdata = new_df, data = lung)
ggsurvplot(fit, conf.int = TRUE, palette = "Dark2", 
           censor = FALSE, surv.median.line = "hv")

This works!
```

**Comment by youyuxiansen** [comment]
> Now, `ggsurvplot()` can be used to plot cox model:
> 
> ```r
> library("survival")
> # Fit cox ph
> #%%%%%%%%%%%%%
> res.cox <- coxph(Surv(time, status) ~ age + sex + wt.loss, data =  lung)
> res.cox
> 
> # Survival curves by sex after adjusting by age and wt.loss
> #%%%%%%%%%%%%%%%%%%%%%%%%
> # we construct a new data frame with two rows, 
> # one for each value of sex; the other covariates are fixed to their average values 
> # Create the new data  
> new_df <- with(lung,
>                data.frame(sex = c(1, 2), 
>                           age = rep(mean(age, na.rm = TRUE), 2),
>                           wt.loss = rep(mean(wt.loss, na.rm = TRUE), 2)
>                )
> )
> new_df
> 
> # Survival curves with new data
> #%%%%%%%%%%%%%%%%%%%%%%%%%%%
> library(survminer)
> fit <- survfit(res.cox, newdata = new_df)
> ggsurvplot(fit, conf.int = TRUE, palette = "Dark2", 
>            censor = FALSE, surv.median.line = "hv")
> ```
> 
> ![rplot](https://cloud.githubusercontent.com/assets/3313355/19835060/f68f8d22-9e7a-11e6-987e-5b1ade7a4eaa.png)

Survival curves plotting in this way, the risk.table just has "ALL", right? @kassambara

**Comment by smmurp2** [comment]
@kassambara I'm having issues getting the CI ribbons to display. Any suggestions?

`cox.tort<-coxph(Surv(Time, Event) ~ Group 
                 + cluster(ID), data = data, na.action = "na.omit")

new.cph.group <- with(data, data.frame(Group = c('resident', 'translocated')))
new.cph.group.fit <- survfit(cox.tort, newdata = new.cph.group)
ggsurvplot(new.cph.group.fit, data = data, conf.int = TRUE, size = 2, censor = FALSE, 
           palette = c("#E69F00", "#56B4E9"), conf.int.style = "ribbon",
           break.time.by = 10, axes.offset = FALSE,
           xlab = "Time (Days)", ylab = "Survival Probability", legend = c(0.2, 0.2), 
           legend.labs = c("Residents", "Translocatees"),
           legend.title = "Group", 
           ggtheme = theme(legend.title = element_text(size = 16, color = "black"),
                 legend.text = element_text(size = 14, color = "black"),
                 axis.text = element_text(size = 14, color = "black"),
                 axis.title = element_text(size = 14, face = "bold", color = "black"),
                 panel.background = element_rect(fill = "white", color = "white"),
                 panel.grid.major = element_line(size = 0.1, linetype = "solid",
                                                 color = "lightgray"))
           )`

Produces the below plot without ribbons:
![image](https://user-images.githubusercontent.com/94133654/141343255-756cd478-56a6-4c16-8189-b8ebd5f36473.png)
